proposals/p2173.md

```diff
@@ -0,0 +1,489 @@
+necessarily transitive: if all equal _pairs_ of types canonicalize to the same
+representation, then _all_ equal types canonicalize to the same representation
+transitively.
+
+### Coherence
+
+In order to avoid surprises in the language behavior, we want Carbon to have the
+property that learning more about the type of a value should not cause the
+behavior of an expression to change from one valid behavior to another valid
+behavior.
+
+It's important to distinguish here between learning more about a type, such as
+learning that it implements a particular interface – for example, from an added
+import – and changing the type, such as by writing a different expression after
+a `:` or `:!`. Changing the type of a value is expected to be able to change the
+behavior of an expression using that value.
+
+## Background
+
+There has been a lot of work in this space. Some previous relevant proposals:
+
+-   [#731: generics details 2](https://github.com/carbon-language/carbon-lang/pull/731)
+    introduced associated types, with the intent that a mechanism to constrain
+    their value would be provided in a later proposal.
+-   [#818: generics details 3](https://github.com/carbon-language/carbon-lang/pull/818)
+    introduced `where` constraints for generics, with an open question for
+    whether and how we might restrict constraints in order to give a decidable
+    type equality rule. #818 addressed that question by using a non-transitive
+    rule with `observe` declarations to close the transitivity gaps.
+-   [#2070: always `==` not `=` in `where` clauses](https://github.com/carbon-language/carbon-lang/pull/2070)
+    describes our previous approach to the "equal types with different
+    interfaces" problem.
+
+## Proposal
+
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
+
+-   `where .A = T` specifies a _rewrite_ constraint. If the associated constant
+    `.A` is named as a member of a type with this constraint, it is rewritten to
+    `T`. This means that the interface of the type changes, and the behavior is
+    in all respects as if `T` had been specified directly.
+-   `where X == Y` specifies a _same-type_ constraint. This is treated much like
+    `where X is SameAs(Y)`. Here, `SameAs` is a hypothetical constraint that is
+    reflexive (`T is SameAs(T)` for all `T`), commutative (`T is SameAs(U)` if
+    and only if `U is SameAs(T)`), and not implementable by the developer.
+    Moreover, if `F` is any pure type function, `T is SameAs(U)` implies
+    `F(T) is SameAs(F(U)`). `observe` statements can be used to form transitive
+    `SameAs` relations. Same-type constraints are not used automatically by the
+    language rules for any purpose, but there is an `ImplicitAs(T)` impl for
+    types implementing `SameAs(T)`.
+
+A type implements `C where .A = T` if and only if it implements
+`C where .A == T` – the inhabitants of these two type-of-types are the same, but
```

Comment:
Technically these two conditions are equivalent only when the first one is legal.